### PR TITLE
initial abstraction layer for communication

### PR DESCRIPTION
**Pull request by fschlimb**

### CODE REVIEW

**Comment by fschlimb** [comment]
Removed hard runtime dependence on MPI
* MPI stuff is now in python module which gets loaded/imported at runtime
* conda recipe adds no runtime dependence to MPI
* you can run non-distributed daal4py without any problem
* when asking for distributed and MPI is not available, an exception is thrown like `ImportError: libmpi.so.12: cannot open shared object file: No such file or directory`
* as a POC you can set D4P_TRANSCEIVER env var; it will try to load that module and bail out if not available: `ModuleNotFoundError: No module named 'no_transceiver'`

@tomashek @triskadecaepyon @oleksandr-pavlyk @shssf

**Comment by fschlimb** [comment]
Lin/OSX python2.7 fail because of new module initialization. Maybe it's time stop supporting python 2.

File: src/mpi/mpi_transceiver.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+public:
+    mpi_transceiver() : transceiver_impl() {}
+
+    virtual void init();
+
+    virtual void fini();
+    
+    virtual size_t nMembers();
+
+    virtual size_t me();
```
Comment (shssf) [inline src/mpi/mpi_transceiver.h]: it would be better to rename `me` to `id`

File: src/mpi/mpi_transceiver.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class mpi_transceiver : public transceiver_impl
+{
+public:
+    mpi_transceiver() : transceiver_impl() {}
+
+    virtual void init();
+
+    virtual void fini();
+    
+    virtual size_t nMembers();
```
Comment (shssf) [inline src/mpi/mpi_transceiver.h]: It would be better to rename `nMembers` to `size`

Comment (fschlimb) [reply src/mpi/mpi_transceiver.h]: size of what? This is not MPI!

File: src/mpi/mpi_transceiver.h (hunk 1 above)
Comment (fschlimb) [reply src/mpi/mpi_transceiver.h]: Not sure I agree. id of what?
Any other opinion?

> It is not clear why so complicated inheritance schema used.

What is unclear? I hoped the comments clarify the intent. I see no other way to provide the same functionality with a clean separation of concerns. Suggestions?

**Comment by fschlimb** [comment]
OK, I'll merge now. We can always suggest changes with new PRs. This is an internal API only, not exposed to users.
At some point someone might want to provide an alternative communication layer, like TCP or OFI. This will be a good time to add the missing implementations in transciever_impl.

**Pull request merged**